platform-sdk/docs/proposals/TSS-Ledger-Id/TSS-Ledger-Id.md

```diff
@@ -0,0 +1,1027 @@
+   1. Sign the ledger ID with the node's private signing key.
+   2. send a `TssVoteTransaction` with the ledger id and the vote vector.
+
+##### Validate TSS Message List
+
+1. For the votes already cast in the vote vector
+   1. sum all the yes votes into a total count.
+   2. If the count is greater than or equal to the threshold in the candidate roster, do nothing and return.
+2. For each TSS message in the TSS Message list
+   1. If the TSS Message has a yes entry in the vote vector, do nothing.
+   2. otherwise, validate the message.
+   3. If the message is valid, update the vote vector and increment the count of yes votes.
+   4. If the count of yes votes is greater than or equal to the threshold in the candidate roster, then send the
+      vote vector as a system transaction and exit the validation process.
+
+#### TSS Key Manager
+
+The TSS Key Manager is responsible for generating the public and private keys of the shares for a fully keyed roster.
+It receives on input the bundle of TssMessages that generate share keys and produces on output the private shares
+for the node, the public shares, and the ledger id.
+
+Internal Elements:
+
+1. Map<RosterHash, Roster> rosters
+2. Map<RosterHash, List<TssMessage>> tssMessages
+3. Map<RosterHash, Record(RosterHash, List<PairingPrivateKey>, List<PairingPublicKey>, LedgerId)> shares
+
+Inputs:
+
+1. Pair<RosterHash, Roster> (Wire)
+2. Pair<RosterHash, List<TssMessage>> (Wire)
+
+Outputs
+
+1. Record(RosterHash, List<PairingPrivateKey>, List<PairingPublicKey>, LedgerId) (Wire)
+
+#### Transaction Resubmitter
+
+The transaction resubmitter is extended to detect which the new `TssMessageTransaction` and `TssVoteTransaction`
+system transactions have not reached consensus because the event containing them has become stale. The Transaction
+Resubmitter will resubmit these transactions to the transaction pool for reprocessing.
+
+### Configuration
+
+The following are new configuration:
+
+1. `tss.maxSharesPerNode` - The maximum number of shares that can be assigned to a node.
+
+### Metrics
+
+The following metrics should be added to the platform:
+
+1. When a candidate roster is set.
+2. The number of TSS Messages collected for a candidate roster.
+3. The number of votes collected for a candidate roster.
+4. The time it takes to compute shares from the key material.
+
+### Performance
+
+The `TSS State Manager` execution on the transaction handling through routes TSS related system transactions to the
+relevant components and updates the TSS related data structures in the state. The routing should be a negligible
+impact provided the receiving components receive the messages and handles them on their own thread. The state
+updates should be simple crud operations, adds and deletes to data structures. All other computation on the
+transaction handling thread should be basic bookkeeping calculations and TSS signature verifications.
```

Comment:
Question: Can signature verification be handled on separate thread(s)?

Signature verification is one of the more expensive things we do, and there is a good amount of work already put into keeping it off the transaction handling thread when possible.